### PR TITLE
Running spark on yarn

### PR DESCRIPTION
**Pull request by mridulm**
Documentation change - I see folks tripping on this in the spark-user group.
Hopefully, this might help things !

### CODE REVIEW

**Comment by mridulm** [comment]
The fixes are done, waiting for review/merge.

File: core/src/main/scala/spark/SparkContext.scala [hunk 1]
```diff
@@ -537,6 +537,8 @@ class SparkContext(
    * filesystems), or an HTTP, HTTPS or FTP URI.
    */
   def addJar(path: String) {
+    // weird - debug why this is happening.
```
Comment (rxin) [inline core/src/main/scala/spark/SparkContext.scala]: Do you have a stack trace on when this happens?

Comment (mridulm) [inline core/src/main/scala/spark/SparkContext.scala]: Not right now (at home)
But it is an NPE coming from addJar usage from SparkContext init ... I will
need to dig into why that is happening.
But that is decoupled from this fix ... Since users can add jar incorrectly
with null as param, it solves that here while allowing examples to run on
yarn.

Regards
Mridul

Comment (mridulm) [inline core/src/main/scala/spark/SparkContext.scala]: Ok, figured out the issue out - SparkPi tries to add a jar from env variable 'SPARK_EXAMPLES_JAR'.
Which need not (or wont) be set when running in yarn mode (we have to use the --jar option).

## So the code snippet from SparkPi

```
val spark = new SparkContext(args(0), "SparkPi",
  System.getenv("SPARK_HOME"), Seq(System.getenv("SPARK_EXAMPLES_JAR")))
```

## 

will cause jars seq to have a null within it : hence the addJar null check is required and solves the issue.

Comment (JoshRosen) [inline core/src/main/scala/spark/SparkContext.scala]: It looks like `SPARK_EXAMPLES_JAR` becoming null has caused problems for other users: https://groups.google.com/d/msg/spark-users/x5UczgI-Xm8/wzMm3Mb77-oJ

File: core/src/main/scala/spark/SparkContext.scala [hunk 2]
```diff
@@ -537,6 +537,8 @@ class SparkContext(
...
+    if (null == path) return
```
Comment (JoshRosen) [inline core/src/main/scala/spark/SparkContext.scala]: I would log a warning warning here, since it will help users to debug their programs if they crash because JARs that they expected to add were not copied to the workers.

File: core/src/main/scala/spark/SparkContext.scala [hunk 3]
```diff
@@ -537,6 +537,10 @@ class SparkContext(
    * filesystems), or an HTTP, HTTPS or FTP URI.
    */
   def addJar(path: String) {
+    if (null == path) {
+      logInfo("null specified as parameter to addJar")
```
Comment (rxin) [inline core/src/main/scala/spark/SparkContext.scala:541]: Can we change this to logWarning, and also print out the stack trace?

Comment (rxin) [inline core/src/main/scala/spark/SparkContext.scala:541]: Actually - do you mind doing this in a separate PR? I am not entirely sure if it is the right behavior to ignore null paths, because it suggests a programming error.

The documentation changes should definitely go in.

Comment (mridulm) [inline core/src/main/scala/spark/SparkContext.scala:541]: The reason this PR was submitted was to fix issues observed by users trying to run spark examples on yarn.

It was initially meant to be only a documentation change : until we observed that SparkPi was making assumptions about environment variables - which are not relevant to yarn mode, but is in other modes.

Actually this api, addJar, is not useful in yarn mode imo.

While it could be argued that all usage of the api (explicitly or implicitly) should defend against this sort of bug, practically (particularly when there are different modes of usage) it is not always possible. Ignore if error is deemed beneign or fail with exception if it is severe - in this case, ignore since it is expected failure.

Ofcourse, this is just my opinion to ensure that exposed api is robust !

Please feel free to cherry pick the changes as required ...

Comment (mridulm) [inline core/src/main/scala/spark/SparkContext.scala:541]: The formatting is getting messed up and resulting in a single blob of text ... sigh.

File: docs/running-on-yarn.md [hunk 4]
```diff
@@ -11,14 +11,32 @@ Ex:  mvn -Phadoop2-yarn clean install
...
   HADOOP_VERSION, HADOOP_MAJOR_VERSION and HADOOP_YARN
 variables before the line 'For Hadoop 2 YARN support'
 Next, uncomment the subsequent 3 variable declaration lines (for these three variables) which enable hadoop yarn support.
 
-Currnetly, it is a TODO to add support for maven assembly.
+Assembly of the jar Ex:  
+./sbt/sbt clean assembly
+
+The assembled jar would typically be something like :
+./streaming/target/spark-streaming-<VERSION>.jar
```
Comment (rxin) [inline docs/running-on-yarn.md]: Why are we using streaming here?

File: core/src/main/scala/spark/SparkContext.scala (hunk 3 above)
Comment (rxin) [inline core/src/main/scala/spark/SparkContext.scala:541]: Thanks. I will do a manual merge. One more issue with the doc update. Can you take a look?

File: docs/running-on-yarn.md (hunk 4 above)
Comment (mridulm) [inline docs/running-on-yarn.md]: That was a typo - I responded with the right path to the users and
copy-pasted wrong path to doc, sigh ...
Good catch - thanks ! Committed it.

On a related note though : the assembly happening in sbt is actually wrong
- ideally we would need to include all aspects of spark in the assembled
  jar - like what is in maven build.
  streaming, repl (others ?) are not included in sbt build ...

Regards,
Mridul

**Comment by rxin** [comment]
I merged this. Thanks!
